### PR TITLE
[SDK] Update StructTag.fromString method to use TypeTagParser

### LINKED ISSUE #7575: [DevEx RFC][TS-SDK] Add support for compound `TypeTag`
## Motivation

Currently, compound / nested type argument is not allowed in Aptos TS-SDK. E.g. 

```
0x5a97986a9d031c4567e15b797be516910cfcb4156312482efc6a19c0a30c948::lp_coin::LP<0xf22bede237a07e121b56d91a491eb7bcdfd1f5907926a9e58338f964a01b17fa::asset::USDC, 0x1::aptos_coin::AptosCoin, 0x190d44266241744264b964a37b8f09863167a12d3e70cda39376cfb4e3561e12::curves::Uncorrelated>
```

The given structure is a LP coin type for Pontem liquidity token. The structure is not able to be called in TS-SDK for coin register / transfer. 

## Proposals

Implement the nested TypeTag for compound structure. 

```
// aptos/src/aptos_types/type_tag.ts
export class StructTag {
    ...
    static fromString(structTag: string): StructTag {
    // Type args are not supported in string literal
    if (structTag.includes("<")) {
      throw new Error("Not implemented");
    }
```

### PR DESCRIPTION
**Pull request by 0xjinn**
### Description
TL:DR -
This is the follow up PR from https://github.com/aptos-labs/aptos-core/pull/7659
Which updates the `fromString()` method and fixes #7575 

Details -
There are two ways of building a transaction and payload
1. `TxnBuilderTypes.TransactionPayloadEntryFunction` - for script transaction
2. `new TransactionBuilderRemoteABI` - for remote abi

The 2nd one is already using the `TypeTagParser`, which is why nested struct tags works, but the 1st one does not directly uses the `TypeTagParser`. If developer decides to use `StructTag.fromString()` - It does not handles nested struct tags. This PR basically updates and make sure the 1st one relies on `TypeTagParser` as well.

### Followup Item(s)
Most of the function from type_tag.ts are not unit tested. Need to increase the coverage of the file.

### Test Plan
1. Manually uses `your_coin.ts` as an example to make sure the `fromString()` works
4. pnpm run lint
5. pnpm test

```
Test Suites: 21 passed, 21 total
Tests:       189 passed, 189 total
Snapshots:   0 total
Time:        142.25 s, estimated 143 s
```

### CODE REVIEW

File: ecosystem/typescript/sdk/src/aptos_types/type_tag.ts [hunk 1]
```diff
@@ -190,17 +190,16 @@ export class StructTag {
    * @returns
    */
   static fromString(structTag: string): StructTag {
-    // Type args are not supported in string literal
-    if (structTag.includes("<")) {
-      throw new Error("Not implemented");
-    }
-
-    const parts = structTag.split("::");
-    if (parts.length !== 3) {
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/aptos_types/type_tag.ts:199]: why remove this check..?

File: ecosystem/typescript/sdk/src/aptos_types/type_tag.ts [hunk 2]
```diff
@@ -190,17 +190,16 @@ export class StructTag {
...
-      throw new Error("Invalid struct tag string literal.");
-    }
-
-    return new StructTag(AccountAddress.fromHex(parts[0]), new Identifier(parts[1]), new Identifier(parts[2]), []);
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/aptos_types/type_tag.ts:203]: we don't need `new Identifier`?

File: ecosystem/typescript/sdk/src/aptos_types/type_tag.ts (hunk 1 above)
Comment (0xjinn) [reply ecosystem/typescript/sdk/src/aptos_types/type_tag.ts:199]: Good question!

We actually want to allow nested struct tag here, so the `length !== 3` check would definitely fail if the input is a nested struct tag. e.g: `0x1::coin::Coin<0x2::coin::MyCoin, 0x1::module::Struct<0x2::coin::MyCoin>>`

File: ecosystem/typescript/sdk/src/aptos_types/type_tag.ts (hunk 2 above)
Comment (0xjinn) [reply ecosystem/typescript/sdk/src/aptos_types/type_tag.ts:203]: Hmm, I think the `module_name` and `name` from line 199-200 are actually Identifier already

**Review by gregnazario — APPROVED** [review]

File: ecosystem/typescript/sdk/src/aptos_types/type_tag.ts [hunk 3]
```diff
@@ -218,3 +217,192 @@ export class StructTag {
     return new StructTag(address, moduleName, name, typeArgs);
   }
 }
+
+function bail(message: string) {
+  throw new Error(message);
```
Comment (davidiw) [inline ecosystem/typescript/sdk/src/aptos_types/type_tag.ts:222]: Let's avoid generic errors and come up with TypeTagParserError or something?

**Review by davidiw — COMMENTED** [review]
cool stuff

File: ecosystem/typescript/sdk/src/aptos_types/type_tag.ts (hunk 3 above)
Comment (0xjinn) [reply ecosystem/typescript/sdk/src/aptos_types/type_tag.ts:222]: Yep sounds good. I can create a new `TypeTagParserError` which extends the generic Error class

File: ecosystem/typescript/sdk/CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,7 @@ All notable changes to the Aptos Node SDK will be captured in this file. This ch
 **Note:** The Aptos TS SDK does not follow semantic version while we are in active development. Instead, breaking changes will be announced with each devnet cut. Once we launch our mainnet, the SDK will follow semantic versioning closely.
 
 ## Unreleased
+- Move `TypeTagParser` from `builder_utils.ts` to `type_tag.ts`
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/CHANGELOG.md:8]: this should have been merged with the previous PR?
also, let's add this PR change into the changelog

File: ecosystem/typescript/sdk/src/aptos_types/type_tag.ts [hunk 5]
```diff
@@ -407,3 +406,10 @@ export class TypeTagParser {
     throw new Error("Invalid type tag.");
   }
 }
+
+class TypeTagParserError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = "TypeTagParserError";
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/aptos_types/type_tag.ts:413]: can we add test showing it throws as expected?

File: ecosystem/typescript/sdk/src/aptos_types/type_tag.ts [hunk 6]
```diff
@@ -407,3 +406,10 @@ export class TypeTagParser {
     throw new Error("Invalid type tag.");
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk/src/aptos_types/type_tag.ts:406]: lets keep it consistent
```suggestion
    bail("Invalid type tag.");
```

**Review by 0xmaayan — APPROVED** [review]
niceeeeeee
